core/txpool/blobpool/blobpool.go

```diff
@@ -0,0 +1,1488 @@
+// draft checks after injection.
+//
+// Note, the method will not initialize the eviction cache values as those will
+// be done once for all transactions belonging to an account after all individual
+// transactions are injected back into the pool.
+func (p *BlobPool) reinject(addr common.Address, tx *types.Transaction) {
+	// Retrieve the associated blob from the limbo. Without the blobs, we cannot
+	// add the transaction back into the pool as it is not mineable.
+	blobs, commits, proofs, err := p.limbo.pull(tx.Hash())
+	if err != nil {
+		log.Error("Blobs unavailable, dropping reorged tx", "err", err)
+		return
+	}
+	// Serialize the transaction back into the primary datastore
+	blob, err := rlp.EncodeToBytes(&blobTx{Tx: tx, Blobs: blobs, Commits: commits, Proofs: proofs})
+	if err != nil {
+		log.Error("Failed to encode transaction for storage", "hash", tx.Hash(), "err", err)
+		return
+	}
+	id, err := p.store.Put(blob)
+	if err != nil {
+		log.Error("Failed to write transaction into storage", "hash", tx.Hash(), "err", err)
+		return
+	}
+	// Update the indixes and metrics
+	meta := newBlobTxMeta(id, p.store.Size(id), tx)
+
+	if _, ok := p.index[addr]; !ok {
+		if err := p.reserve(addr, true); err != nil {
+			log.Warn("Failed to reserve account for blob pool", "tx", tx.Hash(), "from", addr, "err", err)
+			return
+		}
+		p.index[addr] = []*blobTxMeta{meta}
+		p.spent[addr] = meta.costCap
+		p.evict.Push(addr)
+	} else {
+		p.index[addr] = append(p.index[addr], meta)
+		p.spent[addr] = new(uint256.Int).Add(p.spent[addr], meta.costCap)
+	}
+	p.lookup[meta.hash] = meta.id
+	p.stored += uint64(meta.size)
+}
+
+// SetGasTip implements txpool.SubPool, allowing the blob pool's gas requirements
+// to be kept in sync with the main transacion pool's gas requirements.
+func (p *BlobPool) SetGasTip(tip *big.Int) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// Store the new minimum gas tip
+	old := p.gasTip
+	p.gasTip = uint256.MustFromBig(tip)
+
+	// If the min miner fee increased, remove transactions below the new threshold
+	if old == nil || p.gasTip.Cmp(old) > 0 {
+		for addr, txs := range p.index {
+			for i, tx := range txs {
+				if tx.execTipCap.Cmp(p.gasTip) < 0 {
+					// Drop the offending transaction
+					var (
+						ids    = []uint64{tx.id}
+						nonces = []uint64{tx.nonce}
+					)
+					p.spent[addr] = new(uint256.Int).Sub(p.spent[addr], txs[i].costCap)
+					p.stored -= uint64(tx.size)
+					delete(p.lookup, tx.hash)
+					txs[i] = nil
+
+					// Drop everything afterwards, no gaps allowed
+					for j, tx := range txs[i+1:] {
+						ids = append(ids, tx.id)
+						nonces = append(nonces, tx.nonce)
+
+						p.spent[addr] = new(uint256.Int).Sub(p.spent[addr], tx.costCap)
+						p.stored -= uint64(tx.size)
+						delete(p.lookup, tx.hash)
+						txs[i+1+j] = nil
+					}
+					// Clear out the dropped transactions from the index
+					if i > 0 {
+						p.index[addr] = txs[:i]
```

Comment:
Since we are dropping transactions at the end of list, we should also fix the heap.